### PR TITLE
feat(lang): Support env.yaml in conda_packages

### PR DESCRIPTION
**Pull request by sea-wyq**

### CODE REVIEW

**Comment by sea-wyq** [comment]
#671

File: pkg/lang/ir/parser/conda.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	EnvName       string
+	CondaPackages []string
+	PipPackages   []string
+}
+
+func ParseCondaEnvYaml(condaEnvFile string) (*CondaEnv, error) {
+	env := &CondaEnv{}
+	config := viper.New()
+	config.SetConfigFile(condaEnvFile)
+	//尝试进行配置读取
```
Comment (gaocegege) [inline pkg/lang/ir/parser/conda.go:33]: Please use English for the comments, thanks!

File: pkg/lang/ir/parser/conda.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+					env.PipPackages = append(env.CondaPackages, p.(string))
+				}
+			}
+		}
+	}
+	logrus.Debug("parsed python requirements: ", env)
+	return env, nil
+}
+
+//func main() {
```
Comment (gaocegege) [inline pkg/lang/ir/parser/conda.go:54]: Please remove the main here. Maybe you can add some tests if you want to verify if it works.

**Review by gaocegege — COMMENTED** [review]
Thanks for your contribution! :tada: :+1:

Leave some comments here.

File: pkg/lang/ir/parser/conda.go (hunk 1 above)
Comment (sea-wyq) [reply pkg/lang/ir/parser/conda.go:33]: ok

**Comment by gaocegege** [comment]
https://github.com/tensorchord/envd/blob/main/pkg/lang/frontend/starlark/install/install.go#L50

You can refer to the design here to pass a new arg in conda_package

**Comment by aseaday** [comment]
How about use a WIP in your PR title. It will be friendly during a work in progress for other member in the team.

**Comment by sea-wyq** [comment]
image

I modified install.go, but the parameter appeared not found.

File: pkg/lang/frontend/starlark/install/install.go [hunk 3]
```diff
@@ -187,9 +187,10 @@ func ruleFuncVSCode(thread *starlark.Thread, _ *starlark.Builtin,
 func ruleFuncConda(thread *starlark.Thread, _ *starlark.Builtin,
 	args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
 	var name, channel *starlark.List
+	var envFile starlark.String
 
 	if err := starlark.UnpackArgs(ruleConda,
-		args, kwargs, "name", &name, "channel?", &channel); err != nil {
+		args, kwargs, "name", &name, "channel?", &channel, "envFile?", &envFile); err != nil {
```
Comment (gaocegege) [inline pkg/lang/frontend/starlark/install/install.go:193]: How about `env_file` here? We always use lower case in starlark part

**Review by gaocegege — COMMENTED** [review]
Maybe you can share your build.envd to debug it.

File: pkg/lang/frontend/starlark/install/install.go (hunk 3 above)
Comment (sea-wyq) [reply pkg/lang/frontend/starlark/install/install.go:193]: def build():
    install.conda_packages(name = [
        "pytorch",
        "cudatoolkit=11.3",
    ], channel= ["pytorch"])
    base(os="ubuntu20.04", language="python3.8",envFile="env.yaml")
    install.python_packages(name = [
        "flask"
    ])

Comment (sea-wyq) [reply pkg/lang/frontend/starlark/install/install.go:193]: ok ，I try to lowercase

Comment (gaocegege) [reply pkg/lang/frontend/starlark/install/install.go:193]: envFile="env.yaml" should be in `install.conda_packages` instead of `base`

Comment (sea-wyq) [reply pkg/lang/frontend/starlark/install/install.go]: def build():
    install.conda_packages(name = [
        "pytorch",
        "cudatoolkit=11.3",
    ], channel= ["pytorch"],env_file="env.yaml")
    base(os="ubuntu20.04", language="python3.8")
    install.python_packages(name = [
        "flask"
    ])
The above configuration file is wrong

Comment (sea-wyq) [reply pkg/lang/frontend/starlark/install/install.go]: Lowercase parameter names also report an error
image

Comment (gaocegege) [reply pkg/lang/frontend/starlark/install/install.go]: Please use ./bin/envd instead of the envd you installed to run.

Comment (sea-wyq) [reply pkg/lang/frontend/starlark/install/install.go]: ok

Comment (sea-wyq) [reply pkg/lang/frontend/starlark/install/install.go]: #0 0.186 mkdir: cannot create directory '/opt/conda/pkgs': File exists                                                                                          
------
ERRO[2022-07-28T04:33:27Z] failed to load docker image: Post "http://%2Fvar%2Frun%2Fdocker.sock/v1.41/images/load?quiet=1": context canceled  tag="conda:dev"
error: process "mkdir /opt/conda/pkgs" did not complete successfully: exit code: 1

Comment (gaocegege) [reply pkg/lang/frontend/starlark/install/install.go]: Please clean the cache and give another try:

```
./bin/envd prune --all
```

Comment (sea-wyq) [reply pkg/lang/frontend/starlark/install/install.go]: still the same error

Comment (gaocegege) [reply pkg/lang/frontend/starlark/install/install.go]: Could you show the full debug log here?

Comment (sea-wyq) [reply pkg/lang/frontend/starlark/install/install.go]: I added the -p parameter to the mkdir command

Comment (gaocegege) [reply pkg/lang/frontend/starlark/install/install.go]: I did not find where you add it. Could you push the code?

Comment (sea-wyq) [reply pkg/lang/frontend/starlark/install/install.go]: At present, the program can run normally, but the dependencies in the yaml file are not installed

Comment (sea-wyq) [reply pkg/lang/frontend/starlark/install/install.go]: > I did not find where you add it. Could you push the code?

The code has been pushed

Comment (sea-wyq) [reply pkg/lang/frontend/starlark/install/install.go]: yaml file can be downloaded normally

File: examples/conda/build.envd [hunk 4]
```diff
@@ -1,29 +1,9 @@
 def build():
-    config.conda_channel(channel="""
```
Comment (gaocegege) [inline examples/conda/build.envd:2]: Maybe you can avoid changing the code here and create a new example conda-env-yaml for the new feature.

File: pkg/lang/frontend/starlark/install/install.go [hunk 5]
```diff
@@ -206,9 +207,17 @@ func ruleFuncConda(thread *starlark.Thread, _ *starlark.Builtin,
...
+	if env_filestr != "" {
+		buildContextDir := starlark.Universe[builtin.BuildContextDir]
+		buildContextDirStr := buildContextDir.(starlark.String).GoString()
+		buf := filepath.Join(buildContextDirStr, env_filestr)
+		path = &buf
+	}
 
-	logger.Debugf("rule `%s` is invoked, name=%v, channel=%v", ruleConda, nameList, channelList)
-	ir.CondaPackage(nameList, channelList)
+	logger.Debugf("rule `%s` is invoked, name=%v, channel=%v,envFile=%s", ruleConda, nameList, channelList, env_filestr)
```
Comment (gaocegege) [inline pkg/lang/frontend/starlark/install/install.go:219]:
```suggestion
	logger.Debugf("rule `%s` is invoked, name=%v, channel=%v, env_file=%s", ruleConda, nameList, channelList, envFileStr)
```

File: pkg/lang/frontend/starlark/install/install.go [hunk 6]
```diff
@@ -206,9 +207,17 @@ func ruleFuncConda(thread *starlark.Thread, _ *starlark.Builtin,
 			channelList = append(channelList, channel.Index(i).(starlark.String).GoString())
 		}
 	}
+	var path *string = nil
+	env_filestr := env_file.GoString()
```
Comment (gaocegege) [inline pkg/lang/frontend/starlark/install/install.go:211]:
```suggestion
	envFileStr := env_file.GoString()
```

We always use camelcase in golang.

File: pkg/lang/ir/fs.go [hunk 7]
```diff
@@ -19,6 +19,6 @@ import (
 
 func (g Graph) CompileCacheDir(root llb.State, cacheDir string) llb.State {
 	root = llb.User("envd")(root)
-	run := root.Run(llb.Shlexf("mkdir %s", cacheDir), llb.WithCustomName("[internal] create cache dir"))
+	run := root.Run(llb.Shlexf("mkdir -p %s", cacheDir), llb.WithCustomName("[internal] create cache dir"))
```
Comment (gaocegege) [inline pkg/lang/ir/fs.go:22]: I am wondering why we need the -p here. But LGTM.

**Review by gaocegege — COMMENTED** [review]
Thanks for your contribution!

File: examples/conda/build.envd (hunk 4 above)
Comment (sea-wyq) [reply examples/conda/build.envd:2]: ok

File: base-images/python3.8-ubuntu20.04-cuda11.6.Dockerfile [hunk 8]
```diff
@@ -72,7 +72,7 @@ RUN apt-get update && \
 
 RUN apt-get update && \
     apt-get install -y --no-install-recommends --no-install-suggests --fix-missing bash-static \
-    # conda dependencies
+    # conda-yaml dependencies
```
Comment (gaocegege) [inline base-images/python3.8-ubuntu20.04-cuda11.6.Dockerfile:75]: There is no need to update base images, I think.

File: pkg/builder/util.go [hunk 9]
```diff
@@ -63,7 +63,7 @@ func ImageConfigStr(labels map[string]string,
 // DefaultPathEnvUnix is unix style list of directories to search for
 // executables. Each directory is separated from the next by a colon
 // ':' character .
-const DefaultPathEnvUnix = "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/conda/bin:/usr/local/julia/bin"
+const DefaultPathEnvUnix = "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/conda-yaml/bin:/usr/local/julia/bin"
```
Comment (gaocegege) [inline pkg/builder/util.go:66]: Please do not update here.

File: pkg/lang/frontend/starlark/install/install.go [hunk 10]
```diff
@@ -187,9 +187,10 @@ func ruleFuncVSCode(thread *starlark.Thread, _ *starlark.Builtin,
 func ruleFuncConda(thread *starlark.Thread, _ *starlark.Builtin,
 	args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
 	var name, channel *starlark.List
+	var env_file starlark.String
```
Comment (gaocegege) [inline pkg/lang/frontend/starlark/install/install.go:190]: Please use envFile

File: pkg/lang/ir/conda.go [hunk 11]
```diff
@@ -33,10 +33,10 @@ func (g Graph) CondaEnabled() bool {
 
 func (g Graph) compileCondaChannel(root llb.State) llb.State {
 	if g.CondaConfig != nil && g.CondaConfig.CondaChannel != nil {
-		logrus.WithField("conda-channel", *g.CondaChannel).Debug("using custom connda channel")
+		logrus.WithField("conda-yaml-channel", *g.CondaChannel).Debug("using custom connda channel")
```
Comment (gaocegege) [inline pkg/lang/ir/conda.go:36]: Please do not update here.

**Review by gaocegege — COMMENTED** [review]
There is no need to update conda to `conda-yaml` Others LGTM

File: examples/conda-yaml/build.envd [hunk 12]
```diff
@@ -0,0 +1,17 @@
+def build():
+    base(os="ubuntu20.04", language="python3.8")   # 声明基础镜像和conda的python版本
```
Comment (gaocegege) [inline examples/conda-yaml/build.envd:2]: BTW, please use English or remove the comment here.

**Comment by sea-wyq** [comment]
@gaocegege  Fixed

**Review by gaocegege — COMMENTED** [review]
Thanks for your contribution! :tada: :+1:

I know it definitely should be difficult to learn golang and file a feature pull request to envd in only a few days.

Appreciate your work!

File: pkg/lang/frontend/starlark/install/install.go [hunk 13]
```diff
@@ -206,9 +207,17 @@ func ruleFuncConda(thread *starlark.Thread, _ *starlark.Builtin,
...
+		buildContextDir := starlark.Universe[builtin.BuildContextDir]
+		buildContextDirStr := buildContextDir.(starlark.String).GoString()
+		buf := filepath.Join(buildContextDirStr, envFileStr)
+		path = &buf
+	}
 
-	logger.Debugf("rule `%s` is invoked, name=%v, channel=%v", ruleConda, nameList, channelList)
-	ir.CondaPackage(nameList, channelList)
+	logger.Debugf("rule `%s` is invoked, name=%v, channel=%v,envFile=%s", ruleConda, nameList, channelList, envFileStr)
+	ir.CondaPackage(nameList, channelList, path)
```
Comment (gaocegege) [inline pkg/lang/frontend/starlark/install/install.go:220]: Please check if the return value is not nil:

```
if err := ir.CondaPackage(..); err != nil {
  return starlark.None, err
}

File: pkg/lang/ir/parser/conda.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+func ParseCondaEnvYaml(condaEnvFile string) (*CondaEnv, error) {
+	env := &CondaEnv{}
+	config := viper.New()
+	config.SetConfigFile(condaEnvFile)
+	if err := config.ReadInConfig(); err != nil {
+		return nil, errors.Wrapf(err, "failed to read the Conda Env File %s", condaEnvFile)
+	}
+	env.EnvName = config.GetString("name")
+	for _, dependencies := range config.Get("dependencies").([]interface{}) {
+		switch dependencies.(type) {
```
Comment (gaocegege) [inline pkg/lang/ir/parser/conda.go:38]: Please refer to https://blog.csdn.net/u013536232/article/details/109192080 to eliminate type assertions in switch cases

**Comment by gaocegege** [comment]
I am merging the PR although there are some linting issues. I will fix it ASAP.

**Comment by gaocegege** [comment]
@all-contributors Please add @sea-wyq for the code contribution

**Pull request merged**